### PR TITLE
partial support for quantized models in ONNX importer

### PR DESCRIPTION
**Pull request by vpisarev**
See https://github.com/opencv/opencv/issues/20188. Merge together with https://github.com/opencv/opencv_extra/pull/878.

* Note that the code is not fully tested. There is just a smoke test that checks that one simple quantized model is loaded successfully, and then we can run inference on it and get 'some' result of the proper type and size. We need some real quantized deep nets to test the functionality properly.
* QLinearConv is converted to a normal Convolution layer; QLinearMatMul is converted to MatMul. The original 8-bit weights and the quantization/de-quantization factors are preserved though, so Convlution and MatMul layers can, in principle, be extended to support 8-bit weights.
* QuantizeLinear outputs FP32 data; it needs to be extended to output 8-bit tensors.
* DequantizeLinear and QLinearAdd take FP32, as well as INT8/UINT8 inputs, but the output is always FP32 for now. 

The PR is submitted to master branch, rather than 3.4, because the whole support for 8-bit compute paths will be added to master (see #20228)

### Pull Request Readiness Checklist

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.3.0
build_image:Custom Mac=openvino-2021.3.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/all_layers.hpp [hunk 1]
```diff
@@ -689,6 +689,27 @@ CV__DNN_INLINE_NS_BEGIN
...
+        static Ptr<QuantizeLinearLayer> create(const LayerParams& params);
+    };
+
+    class CV_EXPORTS DequantizeLinearLayer : public Layer
+    {
+    public:
+        static Ptr<DequantizeLinearLayer> create(const LayerParams& params);
+    };
+
+    class CV_EXPORTS QLinearAddLayer : public Layer
```
Comment (alalek) [inline modules/dnn/include/opencv2/dnn/all_layers.hpp:704]: It makes sense to add documentation with links on similar layers/nodes (e.g, from ONNX)

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 2]
```diff
@@ -117,7 +117,7 @@ class BaseConvolutionLayerImpl : public ConvolutionLayer
         outputs_arr.getMatVector(outputs);
 
         CV_Assert((inputs.size() > outputs.size() && blobs.empty()) ||
-                  (!inputs.empty() && (blobs.size() == 1 || blobs.size() == 2)));
+                  (!inputs.empty() && blobs.size() >= 1));
```
Comment (alalek) [inline modules/dnn/src/layers/convolution_layer.cpp:120]: what this class should do with blobs.size() = 3 ?

This is assertion check about the code below.
It is caller responsibility to properly call this code without assertions

File: modules/dnn/src/layers/qlinearadd_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+// Copyright (C) 2016, Intel Corporation, all rights reserved.
+// Third party copyrights are property of their respective owners.
```
Comment (alalek) [inline modules/dnn/src/layers/qlinearadd_layer.cpp:6]: probably this is incorrect

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -1286,6 +1289,154 @@ void ONNXImporter::handleNode(const opencv_onnx::NodeProto& node_proto_)
...
+
+            float scale_val[] = {1.f, 1.f, 1.f};
+            int zp_val[] = {0, 0, 0};
+            for(int k = 0; k < 3; k++) {
+                int idx = k == 0 ? 1 : k == 1 ? 4 : 6;
+                Mat scale = blobs[idx], zp = blobs[idx+1];
+                int sc_total = (int)scale.total();
+                if (!scale.empty()) {
+                    if(!((sc_total == 1 || (k == 1 && sc_total == outN)) && scale.type() == CV_32F))
+                        CV_Error(CV_StsError,
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:1328]: > CV_StsError

C API constants should not be used. Here and below.

File: modules/dnn/test/test_onnx_importer.cpp [hunk 5]
```diff
@@ -1339,6 +1339,27 @@ TEST_P(Test_ONNX_nets, Resnet34_kinetics)
...
+{
+    if (backend != DNN_BACKEND_OPENCV || target != DNN_TARGET_CPU)
+        throw SkipTestException("Only the default backend; CPU target is supported");
+
+    String onnxmodel = findDataFile("dnn/onnx/models/model_quant_static.onnx", true);
+    Net net = readNetFromONNX(onnxmodel);
+    ASSERT_FALSE(net.empty());
+    net.setPreferableBackend(backend);
+    net.setPreferableTarget(target);
+    std::cout << net.dump() << "\n";
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:1352]: `if (cvtest::debugLevel > 0)` or remove

File: modules/dnn/test/test_onnx_importer.cpp [hunk 6]
```diff
@@ -1339,6 +1339,27 @@ TEST_P(Test_ONNX_nets, Resnet34_kinetics)
...
+    int shape[] = {1, 3, 16, 16};
+    Mat input(4, shape, CV_32F);
+    RNG& rng = theRNG();
+    rng.fill(input, RNG::UNIFORM, 0, 255);
+    net.setInput(input);
+    Mat output = net.forward();
+    //std::cout << output << std::endl;
+    ASSERT_EQ(output.size(), Size(128, 1));
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:1360]: `*_EQ(expected_value, actual_value)`

File: modules/dnn/test/test_onnx_importer.cpp [hunk 7]
```diff
@@ -1339,6 +1339,27 @@ TEST_P(Test_ONNX_nets, Resnet34_kinetics)
     expectNoFallbacksFromIE(net);
 }
 
+TEST_P(Test_ONNX_nets, QuantStatic)
+{
+    if (backend != DNN_BACKEND_OPENCV || target != DNN_TARGET_CPU)
+        throw SkipTestException("Only the default backend; CPU target is supported");
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:1345]: > default backend

default backend in builds with IE is IE, so this message should be corrected to be accurate.

**Comment by asmorkalov** [comment]
@vpisarev Friendly reminder.

**Pull request closed without merging**